### PR TITLE
Restrict keys to address records in the same canonical name.

### LINKED ISSUE #35: define rules for cname (aka multi-cdn or multi-cloud)
traditional SNI is the origin name. You can use an HTTP connect tunnel, traverse a CNAME (or 3), or engage in some alt-svc load balancing and, while these things all change the DNS name you lookup and connect() to, they don't change the SNI.

what about eSNI? the keys are a property of the server, not really the origin. It seems really hard to coordinate this across administrative domains.

The case I'm particularly worried about is where more than one hosting provider is in the mix.. ISTM a typical way of doing this is LB'ing DNS responses with CNAMEs. (e.g. 50% of www.example.com gets CNAME www.cdnA.com and 50% gets www.cdnB.com).. We've got folks from 2 CDNs as active contributors here who know better than I, but this kind of level of indirection to get you CDN agility seems pretty common.. we would want cdnA and cdnB to be able to each use their own keys.

Even without agility, its a pretty common pattern for an origin to maintain control of their own DNS and only delegate to a CDN under load/whatnot and do that via cname...

that argues to me to allow _esni TXT records on the server lookup, not just the origin name. Perhaps they could override the _esni at the origin scope but that seems like a way to trigger the misconfiguration fallback path.

Also, what about connect tunnels where you do do SNI, but don't traditionally do DNS at all?

### PR DESCRIPTION
**Pull request by mcmanus**
fixes #35 

this would be a bit cleaner if we got rid of the prefix.. which I think we ought to do as well but is separable.

### CODE REVIEW

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -528,6 +530,58 @@ servers which have opted in, thus minimizing compatibility
 issues. However, there are two scenarios where that does not
 apply, as detailed below.
 
+## Load Balancing Multiple Independent Providers {#multi-cdn}
+
+The DNS is used to independently transfer two linked pieces of
+informaton to implement ESNI. The server is identified through DNS
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]:
```suggestion
information to implement ESNI. The server is identified through DNS
```

File: draft-ietf-tls-esni.md [hunk 2]
```diff
@@ -528,6 +530,58 @@ servers which have opted in, thus minimizing compatibility
...
+address records and the server's ESNI configuration is specified in a TXT
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]: does "address records" mean A or AAAA?

File: draft-ietf-tls-esni.md [hunk 3]
```diff
@@ -528,6 +530,58 @@ servers which have opted in, thus minimizing compatibility
...
+     -> CNAME _esni.www.example.com.provider-B.com
+ AAAA ? www.example.com.provider-A.com
+     -> AAAA 2001:DB8::AAAA
+ TXT ? _esni.www.example.com.provider-B.com
+     -> ".. KEY FOR B.."
+~~~~
+
+The client determines the canonical names for the www.example.com TXT
+and AAAA records do not match. It proceeds by obtaining the canonical
+TXT record that matches the canonical address record it already has.
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]: How does it know that the TXT record is the one that needs updating?  It is presented with a mismatch, with TTLs.  Should we instead recommend requesting both, or does that invite problems?

Does it start over from _esni.example.com, or does this allow for an end-run on the CNAME resolution process from the prefixed domain.  That is, if www.example.com is CNAMEd through various intermediaries to providerA.example, does the client just construct _esni.providerA.example, or does it have to go back to _esni.example.com?  I get the problem - if this is load balanced, how do we ensure that this ends up routing the same way?  So why would you not just start with _esni.providerA.example then?

File: draft-ietf-tls-esni.md [hunk 4]
```diff
@@ -528,6 +530,58 @@ servers which have opted in, thus minimizing compatibility
...
+implemented as a DNS CNAME record. This provides a level of indirection
+between the domain owner and the hosting provider. When a domain is
+load balanced, the canonical name of the aliased domain potentially
+changes between different providers on a per-request basis. This can
+lead to the use of uncoordinated addressing and key information from
+different providers which in turn will cause connection failures.
+
+In order to prevent this failure the client confirms that the two
+records being used share the same canonical provenance. Specifically,
+the client MUST confirm that the unprefixed (i.e. lacking the
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]:
```suggestion
the client MUST confirm that the unprefixed (i.e., lacking the
```

File: draft-ietf-tls-esni.md [hunk 5]
```diff
@@ -528,6 +530,58 @@ servers which have opted in, thus minimizing compatibility
...
+
+~~~~
+ AAAA ? www.example.com
+     -> CNAME www.example.com.provider-A.com
+ TXT ? _esni.www.example.com
+     -> CNAME _esni.www.example.com.provider-B.com
+ AAAA ? www.example.com.provider-A.com
+     -> AAAA 2001:DB8::AAAA
+ TXT ? _esni.www.example.com.provider-B.com
+     -> ".. KEY FOR B.."
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]:
```suggestion
     -> "... KEY FOR B ..."
```

File: draft-ietf-tls-esni.md [hunk 6]
```diff
@@ -528,6 +530,58 @@ servers which have opted in, thus minimizing compatibility
...
+~~~~
+
+The client determines the canonical names for the www.example.com TXT
+and AAAA records do not match. It proceeds by obtaining the canonical
+TXT record that matches the canonical address record it already has.
+
+~~~~
+    TXT ? _esni.www.example.com.provider-A.com
+        -> ".. KEY FOR A.."
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md]:
```suggestion
        -> "... KEY FOR A ..."
```

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (mcmanus) [reply draft-ietf-tls-esni.md]: yes.. I don't mind saying a/aaaa I was trying to leave room for other ways of specifying the address. srv? new A* type?

File: draft-ietf-tls-esni.md (hunk 3 above)
Comment (mcmanus) [reply draft-ietf-tls-esni.md]: My thought here is that the TXT record is the new kid on the block, so deference ought to be given to the A for backwards compatibility.

if you're trying to loadbalance, you may well be doing so to add/delete traffic from certain regions (and people have been known to choose providers due to their strengths in particular regions) so capturing the traffic in service of using a valid key seems sub-optimal.. 

I wouldn't be surprised to see keys with bigger TTLs than the A's - creating an easy scenario for the mismatch to occur.. and we wouldn't want the routing of traffic impacted by that imo.

**Comment by mcmanus** [comment]
mostly obe #136 and #137

**Pull request closed without merging**